src/Cognito.re

```diff
@@ -201,6 +206,158 @@ let signUp =
+      | ServerError(_) =>
+        // We're _really_ hoping amazon holds to their API contract here too.
+        // Although the unknownerror variant helps lots.
+        let isErrorResponse = makeResponse(res.json);
+        let err = isErrorResponse->__typeGet;
+        let msg = isErrorResponse->messageGet;
+        let err =
+          switch (err) {
+          | "CodeMismatchException" => `CognitoConfirmationCodeValidation(msg)
+          | _ => `CognitoUnknownError(msg)
+          };
+        Future.make(resolve => resolve(Belt.Result.Error(err)));
+      };
+    });
+};
+
+type signInExceptions = [ | `NotAuthorizedException(string)];
+let signIn = (config, ~username: string, ~password: string, ()) => {
+  let authParams = Js.Dict.empty();
+  Js.Dict.set(authParams, "USERNAME", Js.Json.string(username));
+  Js.Dict.set(authParams, "PASSWORD", Js.Json.string(password));
+
+  let params = Js.Dict.empty();
+  Js.Dict.set(params, "AuthParameters", Js.Json.object_(authParams));
+  Js.Dict.set(params, "AuthFlow", Js.Json.string("USER_PASSWORD_AUTH"));
+  Client.request(config, InitiateAuth, params)
+  ->Future.flatMapOk(res => {
+      /*
+             This is the raw response. Not sure what to do with this if anything. What is `tag:1`
+             Raw Response [ [ 200, tag: 1 ], {} ]
+       Raw Response [ [ 200, tag: 1 ], {} ] */
+      Js.log2("Raw Response", res);
+
+      switch (res.status) {
+      | Success(_) =>
+        // We're _really_ hoping amazon holds to their API contract here.
+        // If not, we're going to see null type errors.
+        let signInResponse = makeSignInResponse(res);
+        let authDecoder = signInResponse->authenticationResultDecoderGet;
+        let authenticationResult = {
+          accessToken: authDecoder->accessTokenGet,
+          idToken: authDecoder->idTokenGet,
+          refreshToken: authDecoder->refreshTokenGet,
+          tokenType: authDecoder->tokenTypeGet,
+          expiresIn: authDecoder->expiresInGet,
+        };
+        Future.make(resolve =>
+          resolve(
+            Belt.Result.Ok({
+              authenticationResult,
+              challengeParameters: signInResponse->challengeParametersGet,
+            }),
+          )
+        );
+      | Informational(_)
+      | Redirect(_)
+      | ClientError(_)
+      | ServerError(_) =>
+        // We're _really_ hoping amazon holds to their API contract here too.
+        // Although the unknownerror variant helps lots.
+        let isErrorResponse = makeResponse(res.json);
+        let err = isErrorResponse->__typeGet;
+        let msg = isErrorResponse->messageGet;
+        let err =
+          switch (err) {
+          | "InvalidParameterException" => `CognitoInvalidParameter(msg)
+          | "NotAuthorizedException" => `CognitoNotAuthorized(msg)
+          | _ => `CognitoUnknownError(msg)
+          };
+        Future.make(resolve => resolve(Belt.Result.Error(err)));
+      };
+    });
+};
+let changePassword =
```

Comment:
Let's delete this entire function and work on the stuff above until it's robust, especially since it looks like it is just copy + paste.

Once ConfirmSignup and SignIn feel like they are in a good mergable state, we can move on.  

Besides, I'm allready seeing a lot of patterns (why I like to colocate early), so there might be an abstraction available that will ***significantly*** reduce the amount of code over a boilerplate approach to building these things.

Reply:
Lead the way, sir.